src/System.Net.HttpListener/src/System/Net/Managed/WebSockets/WebSocketValidate.Http.Managed.cs

```diff
@@ -0,0 +1,112 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    internal static partial class WebSocketValidate
+    {
+        internal static async Task<HttpListenerWebSocketContext> AcceptWebSocketAsyncCore(HttpListenerContext context,
+            string subProtocol,
+            int receiveBufferSize,
+            TimeSpan keepAliveInterval,
+            ArraySegment<byte>? internalBuffer = null)
+        {
+            // get property will create a new response if one doesn't exist.
+            HttpListenerResponse response = context.Response;
+            HttpListenerRequest request = context.Request;
+            ValidateWebSocketHeaders(context);
+
+            string secWebSocketVersion = request.Headers[HttpKnownHeaderNames.SecWebSocketVersion];
+
+            // Optional for non-browser client
+            string origin = request.Headers[HttpKnownHeaderNames.Origin];
+
+            List<string> secWebSocketProtocols = new List<string>();
```

Comment:
This should only construct the list if it's needed.  This can also be moved down to where it's actually used below, e.g.
```C#
List<string> secWebSocketsProtocols = null;
if (shouldSendSecWebSocketProtocolHeader)
{
    secWebSocketsProtocols = new List<string>();
    ...
}
```